crates/aiken-lang/src/tipo/expr.rs

```diff
@@ -1333,26 +1313,23 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
 
     fn infer_clause_pattern(
         &mut self,
-        pattern: UntypedMultiPattern,
-        alternatives: Vec<UntypedMultiPattern>,
-        subjects: &[Arc<Type>],
+        patterns: Vec1<UntypedPattern>,
+        subject: &Type,
         location: &Span,
-    ) -> Result<(TypedMultiPattern, Vec<TypedMultiPattern>), Error> {
+    ) -> Result<Vec<TypedPattern>, Error> {
         let mut pattern_typer = PatternTyper::new(self.environment, &self.hydrator);
 
-        let typed_pattern = pattern_typer.infer_multi_pattern(pattern, subjects, location)?;
-
-        // Each case clause has one or more patterns that may match the
-        // subject in order for the clause to be selected, so we must type
-        // check every pattern.
-        let mut typed_alternatives = Vec::with_capacity(alternatives.len());
-
-        for m in alternatives {
-            typed_alternatives
-                .push(pattern_typer.infer_alternative_multi_pattern(m, subjects, location)?);
+        let mut typed_patterns = Vec::with_capacity(patterns.len());
+        for (ix, pattern) in patterns.into_iter().enumerate() {
+            if ix == 0 {
+                typed_patterns.push(pattern_typer.infer_pattern(pattern, subject)?);
+            } else {
+                typed_patterns
+                    .push(pattern_typer.infer_alternative_pattern(pattern, subject, location)?);
+            }
```

Comment:
This one is a bit odd but I did preserve the original semantic. I don't know where there's a slightly different inference between patterns and "alternative patterns" and I didn't dig too much to be honest. Felt like a rabbit hole.